language/move-prover/interpreter/src/concrete/player.rs

```diff
@@ -0,0 +1,1357 @@
+    }
+    Ok(())
+}
+
+fn handle_move_from(
+    ctxt: &FunctionContext,
+    module_id: ModuleId,
+    struct_id: StructId,
+    ty_args: &[MT::Type],
+    op_addr: RefTypedValue,
+    global_state: &mut GlobalState,
+) -> Result<TypedValue, AbortInfo> {
+    let env = ctxt.target.global_env();
+    let inst = convert_model_struct_type(env, module_id, struct_id, ty_args, &ctxt.ty_args);
+    let addr = op_addr.deref().into_address();
+    match global_state.del_resource(addr, inst) {
+        None => Err(AbortInfo::sys_abort(StatusCode::RESOURCE_DOES_NOT_EXIST)),
+        Some(object) => Ok(object),
+    }
+}
+
+fn handle_get_global(
+    ctxt: &FunctionContext,
+    module_id: ModuleId,
+    struct_id: StructId,
+    ty_args: &[MT::Type],
+    op_addr: RefTypedValue,
+    global_state: &mut GlobalState,
+) -> Result<TypedValue, AbortInfo> {
+    let env = ctxt.target.global_env();
+    let inst = convert_model_struct_type(env, module_id, struct_id, ty_args, &ctxt.ty_args);
+    let addr = op_addr.deref().into_address();
+    match global_state.get_resource(None, addr, inst) {
+        None => Err(AbortInfo::sys_abort(StatusCode::RESOURCE_DOES_NOT_EXIST)),
+        Some(object) => Ok(object),
+    }
+}
+
+fn handle_borrow_global(
+    ctxt: &FunctionContext,
+    module_id: ModuleId,
+    struct_id: StructId,
+    ty_args: &[MT::Type],
+    is_mut: bool,
+    op_addr: RefTypedValue,
+    global_state: &mut GlobalState,
+) -> Result<TypedValue, AbortInfo> {
+    let env = ctxt.target.global_env();
+    let inst = convert_model_struct_type(env, module_id, struct_id, ty_args, &ctxt.ty_args);
+    let addr = op_addr.deref().into_address();
+    match global_state.get_resource(Some(is_mut), addr, inst) {
+        None => Err(AbortInfo::sys_abort(StatusCode::RESOURCE_DOES_NOT_EXIST)),
+        Some(object) => Ok(object),
+    }
+}
+
+fn handle_exists_global(
+    ctxt: &FunctionContext,
+    module_id: ModuleId,
+    struct_id: StructId,
+    ty_args: &[MT::Type],
+    op_addr: RefTypedValue,
+    global_state: &GlobalState,
+) -> TypedValue {
+    let env = ctxt.target.global_env();
+    let inst = convert_model_struct_type(env, module_id, struct_id, ty_args, &ctxt.ty_args);
+    let addr = op_addr.deref().into_address();
+    TypedValue::mk_bool(global_state.has_resource(&addr, &inst))
+}
+
+fn handle_write_back_global_struct(
+    ctxt: &FunctionContext,
+    module_id: ModuleId,
+    struct_id: StructId,
+    ty_args: &[MT::Type],
+    op_struct: RefTypedValue,
+    global_state: &mut GlobalState,
+) {
+    let (struct_ty, object, ptr) = op_struct.deref().decompose();
+    let inst = struct_ty.into_ref_struct_inst(Some(true));
+    if cfg!(debug_assertions) {
+        let env = ctxt.target.global_env();
+        let converted =
+            convert_model_struct_type(env, module_id, struct_id, ty_args, &ctxt.ty_args);
+        assert_eq!(inst, converted);
+    }
+    let addr = match ptr {
+        Pointer::Global(addr) => addr,
```

Comment:
Add a TODO here that this needs to be extended to check for actual address in borrow graph. There is a bug here right now and there is already a TODO in bytcodegenerator (fix is coming up in a while).